designs/component-upgrades.md

```diff
@@ -0,0 +1,149 @@
+
+### Order of execution
+
+Components should be **upgraded** **before the new Kubernetes version is rolled out** (if needed). It’s more common for an old component to have problems with a new Kubernetes version than vice-versa. There might be case by case exceptions, right now there is not, at least with the current set of core components.
+
+This also facilitates CAPI api version upgrades, since updating its components (CRDs and controllers) will also update the current objects in the cluster (through conversion webhooks). This allows us to only maintain only one api version of provider templates, since these are only reapplied when rolling out the new Kubernetes version.
+
+### How to detect new builds for components?
+
+All the CAPI providers objects in the bundles manifest already contain a `version` field. Unfortunately it only reflects the original upstream version that component was built from. We should add the build metadata information to that semver, like `v0.4.2+build-2`. This will allow us to detect new builds as well as preserve the original upstream version, which is important in order to verify api version changes. We might need to trim the build metadata info when interacting with `clusterctl`, I haven’t tested if it’s supported yet.
+
+We will need to add such field to all the other components (Flux, eks-a controllers and crds, cilium...) that don’t have it yet.
+
+### CAPI
+
+All CAPI providers can be updated with `clusterctl`. This works both with version bumps and new builds of the same version. We just need to create the proper overrides layer, configure the images in the clusterctl config file and pass specific versions for every provider to the `upgrade apply` command.
+
+Even if there is  CAPI api version change, the process is the same. The providers will take care of converting the objects in the cluster to the new api version. The only requirement is that all the installed providers need to be upgraded to a version that supports the new api version.
+
+### EKS-A controller and CRDs
+
+Both the controller and CRDs can be upgraded by just applying the new components manifest with kubectl. This should work fine for both new versions and new builds of the same version.
+
+We have to be careful and make sure that any changes to the CRDs are backwards compatible, since at this point there are already eks-a objects in the cluster. Examples of not backwards compatible changes: adding a required field, changing the type of field, making an existing field required, etc. It’s important to note that applying the manifest with the new CRD won’t fail, even if the changes are not backwards compatible. But the objects already in the cluster will become invalid, so any kind of update to those (eg. flux reconciliation) will be be rejected by the kube api server.
```

Comment:
```suggestion
We have to be careful and make sure that any changes to the CRDs are backwards compatible, since at this point there are already EKS Anywhere objects in the cluster. Examples of changes that are not backwards compatible include: adding a required field, changing the type of field, making an existing field required, and so on. It’s important to note that applying the manifest with the new CRD won’t fail, even if the changes are not backwards compatible. But the objects already in the cluster will become invalid. So any kind of update to those (such as flux reconciliation) will be be rejected by the Kubernetes API server.
```